spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -359,31 +359,40 @@ public boolean tryLock(long time, TimeUnit unit) throws InterruptedException {
 		}
 
 		private boolean subscribeLock(long time) throws ExecutionException, InterruptedException {
-			if (!obtainLock()) {
-				if (!RedisLockRegistry.this.redisMessageListenerContainer.isRunning()) {
-					RedisLockRegistry.this.redisMessageListenerContainer.afterPropertiesSet();
-					RedisLockRegistry.this.redisMessageListenerContainer.start();
-				}
+			if (obtainLock()) {
+				return true;
+			}
+
+			if (!RedisLockRegistry.this.redisMessageListenerContainer.isRunning()) {
+				RedisLockRegistry.this.redisMessageListenerContainer.afterPropertiesSet();
+				RedisLockRegistry.this.redisMessageListenerContainer.start();
+			}
+
+			final long expiredTime = System.currentTimeMillis() + time;
```

Comment:
I believe this line has to be a first in the method to calculate the remaining time properly.
The `obtainLock()` and `this.redisMessageListenerContainer.start()` is going to take some time anyway.
So, to be more accurate with expectation we need to `while()` already with the time what was left after those.